contracts/iden3/validators/QueryValidator.sol

```diff
@@ -0,0 +1,136 @@
+ */
+abstract contract QueryValidator is IQueryValidator, OwnableUpgradeable, UUPSUpgradeable {
+    using VerifierHelper for address;
+
+    ILightweightState public override lightweightState;
+    address public override verifier;
+
+    uint256 public override identitesStatesUpdateTime;
+
+    function __QueryValidator_init(
+        address verifierContractAddr_,
+        address stateContractAddr_,
+        uint256 identitesStatesUpdateTime_
+    ) public onlyInitializing {
+        __Ownable_init();
+
+        lightweightState = ILightweightState(stateContractAddr_);
+        verifier = verifierContractAddr_;
+
+        identitesStatesUpdateTime = identitesStatesUpdateTime_;
+    }
+
+    function setVerifier(address newVerifier_) external override onlyOwner {
+        verifier = newVerifier_;
+    }
+
+    function setIdentitesStatesUpdateTime(
+        uint256 newIdentitesStatesUpdateTime_
+    ) public virtual override onlyOwner {
+        identitesStatesUpdateTime = newIdentitesStatesUpdateTime_;
+    }
+
+    function verify(
+        ILightweightState.StatesMerkleData calldata statesMerkleData_,
+        uint256[] memory inputs_,
+        uint256[2] calldata a_,
+        uint256[2][2] calldata b_,
+        uint256[2] calldata c_,
+        uint256 queryHash_
+    ) external view virtual override returns (bool) {
+        require(verifier.verifyProof(inputs_, a_, b_, c_), "QueryValidator: proof is not valid");
+
+        ValidationParams memory validationParams_ = _getInputValidationParameters(inputs_);
+
+        require(
+            validationParams_.queryHash == queryHash_,
+            "QueryValidator: query hash does not match the requested one"
+        );
+
+        require(
+            validationParams_.issuerClaimAuthState == validationParams_.issuerClaimNonRevState,
+            "QueryValidator: only actual states must be used"
+        );
+        require(
+            validationParams_.issuerId == statesMerkleData_.issuerId &&
+                validationParams_.issuerClaimNonRevState == statesMerkleData_.issuerState,
+            "QueryValidator: invalid issuer data in the states merkle data struct"
+        );
+
+        _checkGistRoot(validationParams_.gistRoot);
+        _verifyStatesMerkleData(statesMerkleData_);
+
+        return true;
+    }
+
+    function getCircuitId() external pure virtual returns (string memory);
+
+    function getUserIdIndex() external pure virtual returns (uint256);
+
+    function getChallengeInputIndex() external pure virtual returns (uint256);
+
+    function _getInputValidationParameters(
+        uint256[] memory inputs_
+    ) internal pure virtual returns (ValidationParams memory);
+
+    function _checkGistRoot(uint256 gistRoot_) internal view {
+        ILightweightState.GistRootData memory rootData_ = lightweightState.geGISTRootData(
+            gistRoot_
+        );
+
+        require(
```

Comment:
Can this be frontrun?

Reply:
Nope, as it works for previous roots too (it does not rely on the current root only)